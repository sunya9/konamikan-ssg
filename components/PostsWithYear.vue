<template>
  <div>
    <div v-for="year in years" :key="year" class="content">
      <h2 :id="`#${year}`">
        <span>{{ year }}</span>
        <small class="has-text-grey is-size-6">
          {{ groupedByYear[year].length }} posts
        </small>
      </h2>
      <posts :posts="groupedByYear[year]" />
    </div>
  </div>
</template>
<script lang="ts">
import { Vue, Component, Prop } from 'nuxt-property-decorator'
import { PostOrPage } from '@tryghost/content-api'
import Posts from '~/components/Posts.vue'

@Component({
  components: {
    Posts
  }
})
export default class PostsWithYear extends Vue {
  @Prop({ type: Array, required: true })
  posts!: PostOrPage[]

  get groupedByYear() {
    return this.posts.reduce<{ [key: number]: PostOrPage[] }>((map, post) => {
      if (!post.published_at) return map
      const date = new Date(post.published_at)
      const year = date.getFullYear()
      return {
        ...map,
        [year]: [...(map[year] || []), post]
      }
    }, {})
  }

  get years(): number[] {
    return Object.keys(this.groupedByYear)
      .map((key) => +key)
      .slice()
      .sort((a, b) => b - a)
  }
}
</script>
