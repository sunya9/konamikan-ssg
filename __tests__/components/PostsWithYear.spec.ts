import { mount } from '@vue/test-utils'
import { post } from '@/__tests__/fixtures/postOrPage'
import PostsWithYear from '@/components/PostsWithYear.vue'

describe('PostsWithYear', () => {
  it('normal', () => {
    expect(
      mount(PostsWithYear, {
        propsData: {
          posts: [post]
        }
      })
    ).toMatchSnapshot()
  })
})
