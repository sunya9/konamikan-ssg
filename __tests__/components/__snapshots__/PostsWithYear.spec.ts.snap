// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostsWithYear normal 1`] = `
<div>
  <div class="content">
    <h2 id="#2020"><span>2020</span> <small class="has-text-grey is-size-6">
        1 posts
      </small></h2>
    <ul>
      <li>
        <div class="oneline"><a><span class="has-text-grey-dark">
          04/05
        </span> <span>title</span></a></div>
      </li>
    </ul>
  </div>
</div>
`;
